### DISCUSSION TITLE
Resolving ParsedFunction Scalar Limitation with Volumetric Stress in MOOSE

### DISCUSSION THREAD

**Discussion by AlejandroRiano2023** [idaholab/moose#27591] [Q&A Modules: Phase field] [answered]
Hello, I hope you are doing great.

I'm encountering an issue while trying to activate a hot spot in my simulation based on the volumetric stress field (volstress) being below a certain threshold (0.015 GPa) within a specific region if((x--145.8797)^2+(y-130.4201)^2<(2.7977)^2,1,0). The volstress is calculated as a rank-two scalar auxiliary variable representing the hydrostatic stress.

    [./volstress]
      type = RankTwoScalarAux
      rank_two_tensor = stress
      variable = volstress
      scalar_type = Hydrostatic
    [../]

I've attempted to implement this using a ParsedFunction to define the activation criteria within the [Functions] section of my input file. However, I'm receiving an error indicating that the ParsedFunction only supports scalar variables, and volstress is not scalar.

Here's the relevant portion of my input file:

[Functions]
    [source]
        type = ParsedFunction
        expression = '1e-2*(if(volstress < 0.015, if((x--145.8797)^2 + (y-130.4201)^2 < (2.7977)^2, 1, 0), 0))'
        symbol_names = 'volstress'
        symbol_values = 'volstress'
    []
[]

[Kernels]
    [T_source]
        type = HeatSource
        variable = temperature
        function = source
    []
[]

The error message suggests that ParsedFunction only accepts scalar variables (https://mooseframework.inl.gov/source/functions/MooseParsedFunction.html). However, I need to use the volumetric stress value at each time step to determine when to activate the hot spot.

[31m
*** ERROR ***
The following error occurred in the object "source", of type "ParsedFunction".
The only variables supported by ParsedFunction are scalar variables, and var 'volstress' is not scalar.[39m

How can I modify my approach to effectively use the volumetric stress value within the ParsedFunction while addressing this limitation?

Any advice or suggestions would be greatly appreciated. I'm also willing to provide additional information or code snippets if needed.

Thank you in advance for your assistance.

Best regards,
Alejandro

**Comment by GiudGiud** [comment, marked as answer]
Hello

ParsedFunctions are meant to only depend on x y z and time, and some selected global variables like postprocessors. 

Instead try to use a ParsedMaterial which can depend on variables.
Or a ParsedAux if you are NOT using AD (automatic differentiation)

Guillaume

**Comment by AlejandroRiano2023** [comment]
Hello Guillaume,

Thank you so much for your fast response, I am going to try to implement the code with your suggestions. I will come back telling you how was it.

Best regards,
Alejandro.
